File: frontend/lib/firestore-service.ts
import {
  doc,
  getDoc,
  setDoc,
  updateDoc,
  collection,
  addDoc,
  getDocs,
  query,
  orderBy,
  deleteDoc,
  serverTimestamp
} from 'firebase/firestore';
import { getFirestoreDb } from './firebase';

export interface FirestoreUserProfile {
  firebase_uid: string;
  email: string;
  organization_name: string;
  display_name: string;
  role: string;
  created_at: string;
  updated_at: string;
  auth_provider: 'password' | 'google';
}

export interface SyncProfileOptions {
  organization?: string;
  displayName?: string;
  role?: string;
  authProvider?: 'password' | 'google';
}

export interface ProcurementHistoryRecord {
  procurement_id: string;
  uid: string;
  company_name: string;
  analysis_created_at: string;
  updated_at: string;
  request_status: string;
  request: {
    description?: string;
    requirements: any[];
  };
  items: any[];
  budget: {
    total_budget?: number | null;
    expected_unit_cost?: number | null;
    tolerance?: number | null;
  };
  location: {
    latitude?: number | null;
    longitude?: number | null;
    location_label?: string;
  };
  search: {
    radius?: number | null;
    search_expansion_status?: string;
  };
  sourcing: {
    sourcing_strategy: string;
    official_record_count: number;
    verified_supplier_count: number;
    recommendation_summary: string;
  };
  analysis: {
    readiness?: number;
    standards_matches: any[];
    supplier_results: any[];
    cost_results: any[];
    verification_results: any[];
    response_snapshot: any;
  };
}

/**
 * Synchronizes user profile to Cloud Firestore collection: users/{firebase_uid}
 * - Guaranteed: Uses verified firebase_uid as the document ID.
 * - Guaranteed: Never overwrites existing organization/role on repeat logins.
 * - Guaranteed: Never stores passwords, secrets, or API keys.
 */
export async function syncUserProfileFirestore(
  fbUser: { uid: string; email?: string | null; displayName?: string | null },
  options: SyncProfileOptions = {}
): Promise<boolean> {
  if (typeof window === 'undefined' || !fbUser || !fbUser.uid) {
    return false;
  }

  const db = getFirestoreDb();
  if (!db) {
    return false;
  }

  try {
    const userRef = doc(db, 'users', fbUser.uid);
    const userSnap = await getDoc(userRef);
    const nowIso = new Date().toISOString();

    if (!userSnap.exists()) {
      // First-time user creation
      const profileData: FirestoreUserProfile = {
        firebase_uid: fbUser.uid,
        email: (fbUser.email || '').trim().toLowerCase(),
        organization_name: (options.organization || 'Enterprise Buyer').trim(),
        display_name: (options.displayName || fbUser.displayName || fbUser.email?.split('@')[0] || 'Enterprise User').trim(),
        role: (options.role || 'buyer').trim(),
        created_at: nowIso,
        updated_at: nowIso,
        auth_provider: options.authProvider || 'password'
      };

      await setDoc(userRef, profileData);
      console.log(`[FIRESTORE] Created user document for ${fbUser.uid}`);
    } else {
      // Existing user: safely update auth metadata without overwriting organization or role
      const existingData = userSnap.data();
      const updates: Record<string, any> = {
        updated_at: nowIso
      };

      if (options.authProvider && options.authProvider !== existingData.auth_provider) {
        updates.auth_provider = options.authProvider;
      }
      if (options.displayName && options.displayName !== existingData.display_name) {
        updates.display_name = options.displayName;
      }
      if (options.organization && (!existingData.organization_name || existingData.organization_name === 'Enterprise Buyer')) {
        updates.organization_name = options.organization;
      }

      await updateDoc(userRef, updates);
      console.log(`[FIRESTORE] Updated user document for ${fbUser.uid}`);
    }

    return true;
  } catch (err) {
    console.warn('[FIRESTORE] User profile sync note:', err);
    return false;
  }
}

/**
 * Persists a completed procurement analysis to Cloud Firestore:
 * - procurement_requests/{request_id}
 * - procurement_requests/{request_id}/items/{item_id}
 * - user_activity/{activity_id}
 */
export async function recordProcurementRequestFirestore(
  fbUser: { uid: string },
  requestData: {
    requestId: string;
    organizationName?: string;
    requirements: any;
    normalizedMetadata?: any;
    status?: string;
    items?: Array<{
      item_id: string;
      product_type?: string;
      specifications?: any;
      matched_standards?: any[];
      compliance_status?: string;
    }>;
  }
): Promise<boolean> {
  if (typeof window === 'undefined' || !fbUser || !fbUser.uid || !requestData.requestId) {
    return false;
  }

  const db = getFirestoreDb();
  if (!db) {
    return false;
  }

  try {
    const nowIso = new Date().toISOString();
    const reqRef = doc(db, 'procurement_requests', requestData.requestId);

    const docPayload = {
      firebase_uid: fbUser.uid,
      organization_name: requestData.organizationName || 'Enterprise Buyer',
      procurement_requirements: requestData.requirements,
      normalized_request_metadata: requestData.normalizedMetadata || {},
      status: requestData.status || 'COMPLETED',
      created_at: nowIso,
      updated_at: nowIso
    };

    await setDoc(reqRef, docPayload);

    // Save individual items to subcollection if present
    if (requestData.items && requestData.items.length > 0) {
      for (const item of requestData.items) {
        const itemRef = doc(db, 'procurement_requests', requestData.requestId, 'items', item.item_id);
        await setDoc(itemRef, {
          item_id: item.item_id,
          product_type: item.product_type || 'General Procurement',
          specifications: item.specifications || {},
          matched_standards: item.matched_standards || [],
          compliance_status: item.compliance_status || 'UNKNOWN',
          created_at: nowIso
        });
      }
    }

    // Log lightweight user activity entry
    await recordUserActivityFirestore(fbUser, 'PROCUREMENT_ANALYZED', requestData.requestId);

    console.log(`[FIRESTORE] Recorded procurement request ${requestData.requestId} for ${fbUser.uid}`);
    return true;
  } catch (err) {
    console.warn('[FIRESTORE] Error recording procurement request:', err);
    return false;
  }
}

export async function recordProcurementHistoryFirestore(
  fbUser: { uid: string },
  request: any,
  response: any
): Promise<boolean> {
  if (typeof window === 'undefined' || !fbUser?.uid || !response?.request_id) return false;
  const db = getFirestoreDb();
  if (!db) return false;

  const nowIso = new Date().toISOString();
  const recommendations = response.recommendations || [];
  const record: ProcurementHistoryRecord = {
    procurement_id: response.request_id,
    uid: fbUser.uid,
    company_name: response.company || request.company || 'Enterprise Buyer',
    analysis_created_at: serverTimestamp() as any,
    updated_at: serverTimestamp() as any,
    request_status: 'COMPLETED',
    request: {
      description: request.description,
      requirements: request.requirements || []
    },
    items: (response.items || []).map((item: any) => ({
      item_id: item.item_id,
      item_name: item.item_name,
      quantity: item.normalized_profile?.quantity,
      unit: item.normalized_profile?.unit,
      specifications: item.normalized_profile?.specifications,
      applicable_standards: (item.standards || []).map((standard: any) => standard.is_code),
      compliance_status: item.compliance_status
    })),
    budget: {
      total_budget: request.budget_amount ?? null,
      expected_unit_cost: null,
      tolerance: request.budget_tolerance_pct ?? null
    },
    location: {
      latitude: request.buyer_latitude ?? null,
      longitude: request.buyer_longitude ?? null,
      location_label: response.buyer_location?.city || 'Not provided'
    },
    search: {
      radius: request.search_radius_km ?? null,
      search_expansion_status: response.search_expansion?.message || 'Not expanded'
    },
    sourcing: {
      sourcing_strategy: response.package_sourcing?.strategy || 'ITEM_BY_ITEM_SOURCING',
      official_record_count: (response.official_records || []).length,
      verified_supplier_count: recommendations.filter((item: any) => item.vendor_identity?.gstin_verification?.status === 'VERIFIED').length,
      recommendation_summary: response.package_sourcing?.explanation || ''
    },
    analysis: {
      readiness: response.package_evaluation?.overall_readiness_score,
      standards_matches: (response.items || []).flatMap((item: any) => (item.standards || []).map((standard: any) => standard.is_code)),
      supplier_results: recommendations,
      cost_results: recommendations.map((item: any) => item.cost_assessment),
      verification_results: recommendations.map((item: any) => item.vendor_identity?.gstin_verification),
      response_snapshot: response
    }
  };

  try {
    await setDoc(doc(db, 'users', fbUser.uid, 'procurement_history', response.request_id), record);
    return true;
  } catch (err) {
    console.warn('[FIRESTORE] Procurement history save note:', err);
    return false;
  }
}

export async function listProcurementHistoryFirestore(
  fbUser: { uid: string }
): Promise<{ records: ProcurementHistoryRecord[]; error: boolean }> {
  if (typeof window === 'undefined' || !fbUser?.uid) return { records: [], error: true };
  const db = getFirestoreDb();
  if (!db) return { records: [], error: true };
  try {
    const historyQuery = query(collection(db, 'users', fbUser.uid, 'procurement_history'), orderBy('analysis_created_at', 'desc'));
    const snapshot = await getDocs(historyQuery);
    const records = snapshot.docs.map((item) => {
      const data = item.data() as any;
      const normalizeTimestamp = (value: any): string => {
        if (value?.toDate) return value.toDate().toISOString();
        return typeof value === 'string' ? value : new Date(0).toISOString();
      };
      return {
        ...data,
        analysis_created_at: normalizeTimestamp(data.analysis_created_at),
        updated_at: normalizeTimestamp(data.updated_at)
      } as ProcurementHistoryRecord;
    });
    return { records, error: false };
  } catch (err) {
    console.warn('[FIRESTORE] Procurement history list note:', err);
    return { records: [], error: true };
  }
}

export async function deleteProcurementHistoryFirestore(uid: string, procurementId: string): Promise<boolean> {
  if (typeof window === 'undefined' || !uid || !procurementId) return false;
  const db = getFirestoreDb();
  if (!db) return false;
  try {
    await deleteDoc(doc(db, 'users', uid, 'procurement_history', procurementId));
    return true;
  } catch (err) {
    console.warn('[FIRESTORE] Procurement history delete note:', err);
    return false;
  }
}

/**
 * Records lightweight user activity entry in: user_activity/{activity_id}
 */
export async function recordUserActivityFirestore(
  fbUser: { uid: string },
  action: string,
  requestId?: string
): Promise<boolean> {
  if (typeof window === 'undefined' || !fbUser || !fbUser.uid) {
    return false;
  }

  const db = getFirestoreDb();
  if (!db) {
    return false;
  }

  try {
    const nowIso = new Date().toISOString();
    const activityCollection = collection(db, 'user_activity');

    await addDoc(activityCollection, {
      firebase_uid: fbUser.uid,
      action: action.trim().toUpperCase(),
      request_id: requestId || null,
      timestamp: nowIso
    });

    return true;
  } catch (err) {
    console.warn('[FIRESTORE] User activity recording note:', err);
    return false;
  }
}

File: backend/app/services/firestore_service.py
import os
import uuid
from datetime import datetime, timezone
from typing import Optional, Dict, Any, List
from backend.app.core.config import settings
from backend.app.core.logging import logger
from backend.app.core.firebase import initialize_firebase_admin, FIREBASE_ADMIN_AVAILABLE

try:
    from firebase_admin import firestore
    FIRESTORE_AVAILABLE = True
except ImportError:
    FIRESTORE_AVAILABLE = False


class FirestoreService:
    """
    Cloud Firestore Application Service.
    Manages user profiles, procurement requests, and lightweight activity audit logs.
    
    Security Guarantees:
    - Identity is derived strictly from verified Firebase UID (never client-supplied body).
    - Never stores passwords, secrets, Gemini keys, or database connection strings.
    - Operates with non-blocking graceful fallback in test/offline environments.
    """
    def __init__(self):
        self._db = None
        self._initialized = False

    def _get_client(self):
        if self._initialized:
            return self._db

        self._initialized = True
        if not FIREBASE_ADMIN_AVAILABLE or not FIRESTORE_AVAILABLE:
            logger.debug("[FIRESTORE] firebase-admin firestore package not available.")
            return None

        app = initialize_firebase_admin()
        if not app:
            logger.debug("[FIRESTORE] Firebase Admin app not available; Firestore client disabled.")
            return None

        try:
            self._db = firestore.client(app)
            logger.info(f"[FIRESTORE] Connected to Cloud Firestore for project '{settings.FIREBASE_PROJECT_ID}'")
            return self._db
        except Exception as e:
            logger.debug(f"[FIRESTORE] Firestore client initialization note: {e}")
            return None

    def sync_user_profile(
        self,
        firebase_uid: str,
        email: str,
        display_name: Optional[str] = None,
        organization_name: Optional[str] = None,
        role: str = "buyer",
        auth_provider: str = "password"
    ) -> bool:
        """
        Creates or updates users/{firebase_uid} in Cloud Firestore.
        Preserves user-managed organization and role on repeat logins.
        """
        if not firebase_uid or not firebase_uid.strip():
            return False

        clean_uid = firebase_uid.strip()
        clean_email = email.strip().lower()
        now_iso = datetime.now(timezone.utc).isoformat()

        db = self._get_client()
        if not db:
            return False

        try:
            doc_ref = db.collection("users").document(clean_uid)
            doc_snap = doc_ref.get()

            if not doc_snap.exists:
                doc_payload = {
                    "firebase_uid": clean_uid,
                    "email": clean_email,
                    "organization_name": (organization_name or "Enterprise Buyer").strip(),
                    "display_name": (display_name or clean_email.split("@")[0].title()).strip(),
                    "role": role.strip(),
                    "created_at": now_iso,
                    "updated_at": now_iso,
                    "auth_provider": auth_provider
                }
                doc_ref.set(doc_payload)
                logger.info(f"[FIRESTORE] Created user document in Firestore: users/{clean_uid}")
            else:
                existing = doc_snap.to_dict() or {}
                updates: Dict[str, Any] = {
                    "updated_at": now_iso,
                    "auth_provider": auth_provider
                }
                if display_name and display_name.strip() != existing.get("display_name"):
                    updates["display_name"] = display_name.strip()
                if organization_name and organization_name.strip() and existing.get("organization_name") == "Enterprise Buyer":
                    updates["organization_name"] = organization_name.strip()

                doc_ref.update(updates)
                logger.info(f"[FIRESTORE] Updated user document in Firestore: users/{clean_uid}")

            return True
        except Exception as e:
            logger.warning(f"[FIRESTORE] Note updating user profile in Firestore: {e}")
            return False

    def record_procurement_request(
        self,
        verified_firebase_uid: str,
        request_id: str,
        organization_name: str,
        requirements: Any,
        normalized_metadata: Optional[Dict[str, Any]] = None,
        status: str = "COMPLETED",
        items: Optional[List[Dict[str, Any]]] = None
    ) -> bool:
        """
        Records a procurement request and its items under:
        - procurement_requests/{request_id}
        - procurement_requests/{request_id}/items/{item_id}
        - user_activity/{activity_id}
        """
        if not verified_firebase_uid or not request_id:
            return False

        clean_uid = verified_firebase_uid.strip()
        now_iso = datetime.now(timezone.utc).isoformat()

        db = self._get_client()
        if not db:
            return False

        try:
            # 1. Main request document
            req_ref = db.collection("procurement_requests").document(request_id)
            req_payload = {
                "firebase_uid": clean_uid,
                "organization_name": organization_name.strip(),
                "procurement_requirements": requirements,
                "normalized_request_metadata": normalized_metadata or {},
                "status": status,
                "created_at": now_iso,
                "updated_at": now_iso
            }
            req_ref.set(req_payload)

            # 2. Subcollection items
            if items:
                items_coll = req_ref.collection("items")
                for item in items:
                    item_id = item.get("item_id") or f"item_{uuid.uuid4().hex[:8]}"
                    items_coll.document(item_id).set({
                        "item_id": item_id,
                        "product_type": item.get("product_type", "General Item"),
                        "specifications": item.get("specifications", {}),
                        "matched_standards": item.get("matched_standards", []),
                        "compliance_status": item.get("compliance_status", "UNKNOWN"),
                        "created_at": now_iso
                    })

            # 3. User activity log
            self.record_user_activity(
                verified_firebase_uid=clean_uid,
                action="PROCUREMENT_ANALYZED",
                request_id=request_id
            )

            logger.info(f"[FIRESTORE] Recorded procurement request '{request_id}' for user '{clean_uid}'")
            return True
        except Exception as e:
            logger.warning(f"[FIRESTORE] Error recording procurement request: {e}")
            return False

    def record_user_activity(
        self,
        verified_firebase_uid: str,
        action: str,
        request_id: Optional[str] = None
    ) -> bool:
        """
        Records an append-only activity log in user_activity/{activity_id}.
        """
        if not verified_firebase_uid:
            return False

        clean_uid = verified_firebase_uid.strip()
        now_iso = datetime.now(timezone.utc).isoformat()

        db = self._get_client()
        if not db:
            return False

        try:
            activity_id = f"act_{uuid.uuid4().hex}"
            db.collection("user_activity").document(activity_id).set({
                "activity_id": activity_id,
                "firebase_uid": clean_uid,
                "action": action.strip().toUpperCase() if hasattr(action, "toUpperCase") else action.strip().upper(),
                "request_id": request_id,
                "timestamp": now_iso
            })
            return True
        except Exception as e:
            logger.debug(f"[FIRESTORE] Note recording user activity: {e}")
            return False


# Singleton instance
firestore_service = FirestoreService()
